xarray/core/dataset.py

```diff
@@ -7972,7 +7972,7 @@ def polyfit(
                 scale_da = scale
 
             if w is not None:
-                rhs *= w[:, np.newaxis]
+                rhs = rhs * w[:, np.newaxis]
```

Comment:
How about `rhs = da.copy()` in the `else` above? That would preserve the optimisation for the other code path

Reply:
I tried it, but it's not enough. The test I added fails. We'd have to do this:
```python
            if dims_to_stack:
                stacked_dim = utils.get_temp_dimname(dims_to_stack, "stacked")
                rhs = da.transpose(dim, *dims_to_stack).stack(
                    {stacked_dim: dims_to_stack}
                )
                stacked_coords = {stacked_dim: rhs[stacked_dim]}
                scale_da = scale[:, np.newaxis]
            else:
                rhs = da
                scale_da = scale
            rhs = rhs.copy()
```
Do you think that's better than changin `*=`?

Reply:
I guess it would be worst as we copy even if `w is None`

Reply:
Oh hmm.. OK.